Map wrap

DajanaSnopkova opened the pull request:
First attempt with L.vectorGrid. This is not a working solution. The following things are not resolved:
- stateStyle function - fillColor is not interpreted correctly per each feature in topo json
- statesLayer.filter() does not work - initially 50m layer is loaded, after sufficient zoom also 10m layer loads, but 50m layer is still present 
- addLayerInteraction() still needs to be rewritten to work for L.vectorGrid layer

SLeitgeb committed:
Use TopoJSON as data files

Resolves #2 and #6. Rebuilding data is more robust than loading the
required data asynchronously.

SLeitgeb committed:
Load higher resolution data at zoom level 5

Resolves #3.

SLeitgeb committed:
Refactor TopoJSON parsing

Clean up workflow by moving TopoJSON parsing to a dedicated function.

DajanaSnopkova committed:
init attempt with vectorGrid layer

SLeitgeb commented on map.js:
```diff
@@ -46,51 +46,90 @@ legend.onAdd = function() {
+            return stateStyle(properties);
+          }
+        }
+      }).addTo(map);
```
You are adding the layer directly to the map, not to `statesLayer`. The `.min_zoom` property is also no longer specified for the layers, so the zoom restriction wouldn't work anyway.

DajanaSnopkova replied:
layers are now added to the statesLayer. Zooming issue is fixed.

SLeitgeb commented on map.js:
```diff
@@ -46,51 +46,90 @@ legend.onAdd = function() {
-  style: stateStyle
-});
-statesLayer.addLayer(states50m);
+let states50m = '';
```
You can try defining the layer as an empty `L.vectorGrid`, but you'd have to find out how to add the data to the layer later (not redefining it with `L.vectorGrid.slicer`.

Or we can probably drop the global definition altogether and just instantiate the layer directly in the `fetch.then` method.

SLeitgeb replied:
You would also need to get the `states50m` object reference back from the `statesLayer` group for the call on line 115, so I think it's ok to keep the global reference (but it's better to define it without value like this: `let states50m;`.

DajanaSnopkova replied:
Fixed. `let states50m;` declaration used.

SLeitgeb commented on map.js:
```diff
@@ -46,51 +46,90 @@ legend.onAdd = function() {
-function stateStyle(feature) {
+function stateStyle(properties) {
   // Set the style based on the number of students
+  var d = properties[studentsCountAttr];
```
No `var` plox.

DajanaSnopkova replied:
Fixed.

SLeitgeb commented on map.js:
```diff
@@ -46,51 +46,90 @@ legend.onAdd = function() {
+        states10m = L.vectorGrid.slicer(json, {
+          rendererFactory: L.canvas.tile,
+          vectorTileLayerStyles: {
+            countries: function(properties) {
```
You don't need to set the `countries` property to an anonymous function, you can set it to the `stateStyle` function directly:
```js
vectorTileLayerStyles: {
  countries: stateStyle
}
```

DajanaSnopkova replied:
I need to work with "properties" of the layer, so I don't think this notation would work... I proposed something different, but the Styling issue is still not resolved.

DajanaSnopkova replied:
set to `countries: stateStyle` as proposed

SLeitgeb reviewed (changes requested):
In summary, you need to add the `vectorGrid` layers to the `statesLayer` group for the zoom constrains to work.

DajanaSnopkova committed:
resolve zooming  and formatting issues

DajanaSnopkova committed:
fix color styling

SLeitgeb commented on map.js:
```diff
@@ -87,11 +87,12 @@ fetch(states50mSource)
-      }).addTo(map);
+      });
+  })
+  .then(() => {
```
There's no need for this extra `.then` method, it's more readable to assign to `states50m` and call `statesLayer.addLayer(states50m)` in the same `.then` block.

There was an extra block before, because I wanted to use `loadTopoToLayer` as callback like this:
```js
fetch(…)
  …
  .then(loadTopoToLayer)
  .then(() => {
    states50m.max_zoom = ZOOM_THRESHOLD;
    statesLayer.filter();
  });
```

But I needed to call the `loadTopoToLayer` explicitly in the end anyway and forgot to combine the `.then` methods, it would ideally look like this:
```js
fetch(…)
  …
  .then(topology => {
    loadTopoToLayer(topology, states10m);
    states50m.max_zoom = ZOOM_THRESHOLD;
    statesLayer.filter();
  });
```

Try using just two callbacks per `fetch` if possible:
```js
fetch(states10mSource)
  .then(response => response.json())
  .then(topology => {…});
```

DajanaSnopkova replied:
extra `.then` method removed.

SLeitgeb commented on map.js:
```diff
@@ -1,5 +1,5 @@
 /* eslint-disable space-before-function-paren */
-/* global L topojson ZoomShowHide */
```
Don't forget to also remove the `topojson` import from `index.html`.

DajanaSnopkova replied:
removed

SLeitgeb commented on map.js:
```diff
@@ -104,24 +105,23 @@ function zoomHandler() {
-    states50m.max_zoom = ZOOM_THRESHOLD;
-    statesLayer.filter();
-  });
+    .then(() => {
```
Dtto

DajanaSnopkova replied:
extra `.then` method removed

SLeitgeb commented on map.js:
```diff
@@ -131,6 +131,7 @@ function stateStyle(properties) {
                d > 0   ? '#00af3f' :
                          '#C6C6C6',
     color: 'white',
+    fill: true,
```
Just a minor thing, it might look better if we group all fill-related properties in the return statement together, e.g.
```js
{
  fill,
  fillColor,
  fillOpacity,
  color,
  weight
}
```

DajanaSnopkova replied:
updated

DajanaSnopkova committed:
resolve code-review issues

DajanaSnopkova committed:
add IDs to topojsons

DajanaSnopkova committed:
attempt to resolve interaction with vectorGrid layer

SLeitgeb committed:
Include an ID in the TopoJSON features

SLeitgeb committed:
Switch to the SVG render engine for performance, lint

SLeitgeb committed:
Fix tooltip display

SLeitgeb committed:
Bring all parts of the hovered feature to front

A bit of an ugly hack. The feature is sliced to multiple vector tiles
when the layer is loaded, all parts need to be brought forth, otherwise
just the part of the currently hovered tile is made more visible.

There are still issues with eastern Russia and the Alaskan part of USA
at lower zoom levels.

SLeitgeb commented:
The `L.vectorGrid` solution seems to have too many issues, closing for now.
